ad/config.go

```diff
@@ -10,15 +10,27 @@ import (
 type Config struct {
 	Domain   string
 	IP       string
+	URL      string
 	Username string
 	Password string
 }
 
 // Client() returns a connection for accessing AD services.
 func (c *Config) Client() (*ldap.Conn, error) {
 	var username string
+	var url string
 	username = c.Username + "@" + c.Domain
-	adConn, err := clientConnect(c.IP, username, c.Password)
+
+	// stay downwards compatible
+	if c.IP != "" {
+		url = fmt.Sprintf("ldap://%s:389", c.IP)
+	} else if c.URL != "" {
+		url = c.URL
+	} else {
+		return nil, fmt.Errorf("Need either an IP or LDAP URL to connect to AD, check provider configuration")
+	}
```

Comment:
```suggestion
	switch {
	case c.URL:
		url = c.URL
	case c.IP:
		url = fmt.Sprintf("ldap://%s:389", c.IP)
	default:
		return nil, fmt.Errorf("Need either an IP or LDAP URL to connect to AD, check provider configuration")
	}
```

- Ich bevorzuge bei sowas `switch{}`, weil ich es besser lesbar finde.
- Check auf `!= ""` würde den `nil` Fall ggf. nicht abfangen, also lieber auf `true` checken
- Check auf `c.URL` zuerst, um den neuen Parameter zu bevorzugen, wfalls beide gesetzt sind
- Style: Eigentlich sollte in solchen Funktionen nicht geloggt werden, das sollte eigentlich nur im Provider Package passieren. Falls man das so machen würde, ließe sich das ganze deutlich verkürzen, indem man den Error einfach direkt returned:

```suggestion
	switch {
	case c.URL:
		return clientConnect(c.URL, username, c.Password)
	case c.IP:
		return clientConnect(fmt.Sprintf("ldap://%s:389", c.IP), username, c.Password)
	default:
		return nil, fmt.Errorf("Need either an IP or LDAP URL to connect to AD, check provider configuration")
	}
```

Dann könnte man die Funktion direkt beenden und bräuchte auch die Variable `url` nicht....

Reply:
das geht so nicht `invalid case c.URL in switch (mismatched types string and bool)`